### ISSUE TITLE
Does FN really help?

### ISSUE THREAD

**Issue by happynear** [KaleidoZhouYN/Details-on-Face-Recognition#1]
As far as I know, the FN is actually Batch Normalization applied after final feature layer. I used to do such experiments about one year ago because we usually do mean-subtraction on LFW's training set. But I didn't see any benefits of it. Are you sure that FN really helps network to converge? I will consider to use it in the future...

**Comment by KaleidoZhouYN** [comment]
@happynear
This question is quite hard to answer because BatchNorm is beyong my understanding and the "deep visage" paper didn't expain why this method useful either.
But for the ResFace Network it works.If only use softmax,you should aritifitially choose weight decay or the loss will become very small and acurracy on training set will become almost 1 which is definetly overfit.
The same as L2-constrain,or NormFace, FN is also a type of feature length constrain.But no that directly as L2-constrain.FN constrain the distribution of feature,which indirectly constrained the feature length.But it's hard to prove it. 
The accuracy on LFW won't be too much different whether you use FN method if you are using a strong constrained loss like A-Softmax,but it did help converging if you are using deeper network like 64-layers resnet.
One drawback of FN is that you need to use quite large batch size to train the network.so this method is not suited for very deep network like inception-resnet if your GPU memory is not enough.

**Comment by happynear** [comment]
I can train 64-layer resnet using the default settings without difficulty. Is training on larger datasets such as MS-Celeb difficult?

**Comment by KaleidoZhouYN** [comment]
Won't be difficult but if you want to train by A-Softmax this will be some trouble when you change the type of "SINGLE" to "DOUBLE",it's hard to find a property lambda.While using FN,this will be okay.

**Comment by zhouhui1992** [comment]
why did you set the type to 'DOUBLE' rather than 'QUADRUPLE',this maybe get better accuracy.

**Comment by KaleidoZhouYN** [comment]
@zhouhui1992  you can set type to 'QUADRUPLE' when your feature dim is 512(maybe,I'm not sure),but we use 2048 and find 'DOUBLE' can get the best accuracy.

**Comment by vzhangmeng726** [comment]
@KaleidoZhouYN  Thanks for sharing your work, Batchnorm and Scale layers are commonly used together in caffe.  However, there is no scale layer in you provided deploy file。Do you use scale layer? Thank you！

**Comment by vzhangmeng726** [comment]
@KaleidoZhouYN  why you set feature dim is 2048, The general face model is set to 512 dimensions，Generally, it's set to 512 dimensions，Even if you refer to the paper(DeepVisage: Making face recognition simple yet with powerful generalization skills). So, can you explain it in detail? Thank you！

**Comment by KaleidoZhouYN** [comment]
@vzhangmeng726  
1.you can add scale layer with scale = 1 and lr = 0.
2.2048 is more suitable for our own dataset.you can change it to 512.

**Comment by wy1iu** [comment]
With proper setting, BatchNorm can help SphereFace-64 converge easily on MS-1M.

**Comment by IrvingShu** [comment]
Can you tell me that how is the ACC when you use orgin softmax on your MS-1M.

**Comment by nyyznyyz1991** [comment]
@KaleidoZhouYN  I used FN to train, but the saved variance is NAN. In inference phase, the embedding feature( use_global_status = true) is used , so all the feature dimension is zero.  have you ever run into this problem? Maybe my dataset is not cleaned thoroughly?

**Comment by KaleidoZhouYN** [comment]
@nyyznyyz1991 
1.use default paraments of BatchNorm.
2.try again and test the first checkpoint to see whether it is NAN
3.try larger batch_size or try smaller lr like 0.01 at first,after you make sure it is converging,change the lr back to 0.1

**Comment by nyyznyyz1991** [comment]
@KaleidoZhouYN 
Thanks a lot, i have trained a 99.5% model using fn and sphereloss, there are some confusing questions:
1. using fn layer would save the trained mean and variance, which makes the trained model much larger than normal modal (in my case 200M -> 480M).
2. when i design the decay function of lambda(parameter between exp(cos(m*theta)) and exp(cos(theta)) ) and utilze it to train my model, the loss will increase, but the accuracy in lfw doesn't decrease. 
maybe when lambda is small, sphereloss dominated the  mixed loss and sphereloss is in fact much larger than norm softmax loss.
3. how did you choose which iter's caffemodel to restore in step3, and would you give some tips on designing lambda's decay function?

**Comment by KaleidoZhouYN** [comment]
@nyyznyyz1991 
1.BN will add at most 5 parameters in each dimension(in our case it is 5*2048),so i think there will be other reasons for you model space growth.
2.yes,what you say is right
3.normally i change the iter when accuracy is stable.lambda's decay function is different for different dataset,but always keeping in mind your trainning should do classification first then do angle margin.

**Comment by YongtaoGE** [comment]
@KaleidoZhouYN 
1.Could you explain why FN  requires large batch size to train the network? 
2.To what extend the batch size will be different using FN or not?
